### ISSUE TITLE
ComponentBase.BuildRenderTree cannot be overriden but I need to override, unless there's alternate

### ISSUE THREAD

**Issue by htmlsplash** [dotnet/aspnetcore#51140]
## Background and Motivation

I would like to override "ComponentBase.BuildRenderTree" method so I can execute it under specific context.

##Reason for Request:

I need to setup some ambient data before calling BuildRenderTree() method (I follow this pattern outlined here:
https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/dependency-injection?view=aspnetcore-7.0, scroll to the heading: "Access server-side Blazor services from a different DI scope"). My requirement is that I may need to access the database while rendering the component and for this I have to setup hibernate session as ambient data.

##More Background:
We're converting our application from Webforms to Blazor and access to hibernate session was shared for the current HTTP request (stored in current HTTP context). I need to keep the same API design, so we're using ambient data (via AsyncLocal variable).

The code below describes the intent. We want to execute any rendering of markup with razor expressions under active hibernate session context. 

With my current solution  (which allows sharing of ambient data across any method I override in the component "as copied from the pattern: Access server-side Blazor services from a different DI scope" but without overriding BuildRenderTree method) I could populate everything into variables (in the @code section) and then render the variables in markup but that's a workaround that requires unnecessary code.

My code:
FYI: IHibernateSessionAccessor SessionAccessor  - internally uses AsyncLocal variable

`
[Inject]
private IHibernateSessionAccessor SessionAccessor { get; set; }

protected override void BuildRenderTree(RenderTreeBuilder builder)
	=> _InvokeWithHibernateSessionContext(() => base.BuildRenderTree(builder));

private void _InvokeWithHibernateSessionContext(Action func)
{
	_EnsureSession();
	try
	{
		SessionAccessor.Session = _session;
		func();
	}
	finally
	{
		SessionAccessor.Session = null;
	}
}

private ISession _EnsureSession()
{
	if (_session is null)
	{
		_session = HibernateContext.Current.SessionFactory.OpenSession();
	}
	return _session;
}
`

## Proposed API

Please allow developers to override ComponentBase.BuildRenderTree() method or provide an alternate way setup ambient context before rending component's markup.

**Comment by SteveSandersonMS** [comment]
Can you clarify more what's stopping you from doing this already?

It looks like you are simply performing a synchronous operation. Why not do this in `OnInitialized` or even in the component's constructor? The initial render won't happen until `OnInitialized` is synchronously completed.

**Comment by htmlsplash** [comment]
@SteveSandersonMS 
Sure thing, couple scenarios:
Reminder:
We're porting existing Webforms app to Blazor server app.

Scenario 1: Consider the following markup in a Blazor component

@loc.GetResourceString("Label_Password"): <input />

The variable loc is our custom built object (injected via DI) for looking up localized strings (it is custom built because we have additional rules when looking up localizable strings). The line above just simply looks language specific string for resource "Label_Password". But internally, the object applies some specific rules during lookup and one of the rules requires a database query to look up the application settings. This query uses hibernate session retrieved from current HTTP context (ambient context).

I understand, I can have workarounds to make it work, but that requires more effort + rewriting/redesign of some features.

Scenario 2: 
In webforms we used to have declarative way of connecting data to controls via data source controls. At run time the data source control would perform query:

Ex code:
<asp:ObjectDataSource ID="dsRoles" runat="server"
				TypeName="PersistentUserRole"
				SelectMethod="LoadAll"
				DataObjectTypeName="Entity.UserRole"></asp:ObjectDataSource>
<asp:DropDownList ID="ddlRoles" runat="server"
				DataSourceID="dsRoles"
				DataTextField="Description" DataValueField="Id" />

We probably will not be using object data source controls in the new version just because there are more draw backs then benefits of using them but it a valid use case for others?

Conclusion:
I only listed 2 scenarios that come to mind based on our current conversion process, but there maybe others. 

In the webforms paradigm, there's so much use/abuse of ambient context (ie data) that might be accessed in markup code. Sometimes, it is easy to change things, other times it requires more effort and significant re-design which preferably we don't want to do at this moment, but maybe later.

Also, another hurdle is that since we're using the strangler fig approach to convert, some of the data (ex Application settings) is modified in the webforms app, caching this data right now in the Blazor app is not an option ... we have to always go to the database to query it since it may be changed by the admin at any time (in the webforms app). This means, that using application settings always requires a database and always requires a hibernate session to be present in the ambient context.

I am always up for using the code section for loading required data that is bound to the controls (it is the Blazor way of doing things, yields better design and is obvious to the reader) but sometimes some of our legacy apis that are directly used in markup require ambient data.

**Comment by SteveSandersonMS** [comment]
Thanks, but can you clarify what part of that can't be done by populating whatever data you need in `OnInitialized`? It looks to me (and sorry if I'm being naive) that all of it would be doable that way.

**Comment by SteveSandersonMS** [comment]
> It sounds like there's a compelling reason not to allow user to override BuildRenderTree() method, can you share why this is so? In the past, wasn't this method overridable?

It is overridable and always has been. However, the Razor compiler generates code for that override. If you subclass the component type you can add a further layer of override in C# if you wish.

> We have another use case where we need access to the database while rendering markup.
Our Menu control that is present on every page has its menu options trimmed based on currently logged in user access rights. For this to happen we are required to load the user record from the database. Therefore, when rendering the menu, a hibernate session is required to be able to load the user data and perform the trimming.

Sorry, I still don't understand why you think this means you have to access the database while in the middle of rendering. What stops you from loading the user data and determining the set of menu options during `OnInitializedAsync`?

**Comment by htmlsplash** [comment]
"Sorry, I still don't understand why you think this means you have to access the database while in the middle of rendering. What stops you from loading the user data and determining the set of menu options during OnInitializedAsync?" - I think, this might be my lack of knowledge of Blazor. I got around the problem with the Menu with little help, by using OnInitialized() to load all the data, wrapped it around one class, and passed it to my Menu as a parameter. When the menu renders, it uses the parameter (and all the requires values it wraps) to trim the nodes.

"It is overridable and always has been. However, the Razor compiler generates code for that override. If you subclass the component type you can add a further layer of override in C# if you wish." - I'll try that. I wonder if having partial methods in C# (similar to partial classes) would solve this problem. It is inconvenient (and feels maybe a bit hackish) to have to sub-class to make something work, or get around some tool.

**Comment by htmlsplash** [comment]
" If you subclass the component type you can add a further layer of override in C# if you wish." - I tried this and it did not work.

In my example I have: "Test1.razor"

I added the following code to override "BuildRenderTree".

`public class Test1ComponentOverride : Test1
{
	protected override void BuildRenderTree(RenderTreeBuilder __builder)
	{
		base.BuildRenderTree(__builder);
	}
}`

My override is not called.
FYI: Test1.razor component is marked with "page" directive so that it is routable. Is that maybe the cause of the problem?

**Comment by SteveSandersonMS** [comment]
> FYI: Test1.razor component is marked with "page" directive so that it is routable. Is that maybe the cause of the problem?

Yes, you need to remove that and put a `[RouteAttribute]` on `Test1ComponentOverride` instead.
